autopilot/followpoint.cpp

```diff
@@ -0,0 +1,145 @@
+            qDebug() << "WARNING: Follow point timed out. Stopping WaypointFollower.";
+            this->stopFollowPoint();
+        }
+        mFollowPointTimedOut = true;
+    });
+}
+
+FollowPoint::FollowPoint(QSharedPointer<VehicleConnection> vehicleConnection, PosType posTypeUsed)
+{
+    mVehicleConnection = vehicleConnection;
+    connect(&mUpdateStateTimer, &QTimer::timeout, this, &FollowPoint::updateState);
+    setPosTypeUsed(posTypeUsed);
+
+    // TODO: follow point over remote connection not supported for now
+}
+
+bool FollowPoint::isActive()
+{
+    return mUpdateStateTimer.isActive();
+}
+
+void FollowPoint::startFollowPoint()
+{
+    // Deactivate emergency brake in order to use follow point
+    emit deactivateEmergencyBrake();
+    if (isOnVehicle()) {
+        mStmState = FollowPointSTMstates::FOLLOWING;
+        mUpdateStateTimer.start(mUpdateStatePeriod_ms);
+    } else {
+        if (isOnVehicle())
+            qDebug() << "WARNING: Follow Point did not get a recent point to follow. Exiting.";
+        else
+            qDebug() << "WARNING: Follow Point not implemented for remote connections. Exiting.";
+        mStmState = FollowPointSTMstates::STOP;
+    }
+}
+
+void FollowPoint::stopFollowPoint()
+{
+    mUpdateStateTimer.stop();
+    (isOnVehicle() ? mMovementController->getVehicleState() : mVehicleConnection->getVehicleState())->setAutopilotRadius(0);
+    holdPosition();
+    mStmState = FollowPointSTMstates::STOP;
+    emit activateEmergencyBrake();
+}
+
+void FollowPoint::holdPosition()
+{
+    if (isOnVehicle()) {
+        mMovementController->setDesiredSteering(0.0);
+        mMovementController->setDesiredSpeed(0.0);
+    } else {
+        mVehicleConnection->requestVelocityAndYaw({}, mVehicleConnection->getVehicleState()->getPosition(mPosTypeUsed).getYaw());
+    }
+}
+
+void FollowPoint::setPosTypeUsed(const PosType &posTypeUsed)
+{
+    mPosTypeUsed = posTypeUsed;
+}
+
+void FollowPoint::updateFollowPointInVehicleFrame(const PosPoint &point)
+{
+    if ((mStmState == FollowPointSTMstates::FOLLOWING || mStmState == FollowPointSTMstates::WAITING) &&
+         (point.getTime() > mCurrentFollowPointInVehicleFrame.getTime())) {
+        if (mFollowPointTimedOut) {
+            qDebug() << "Follow Point: timeout reset.";
+        }
+
+        mCurrentFollowPointInVehicleFrame = point;
+        mFollowPointHeartbeatTimer.start(mFollowPointTimeout_ms);
+        mFollowPointTimedOut = false;
+    }
+}
+
+PosPoint FollowPoint::getCurrentVehiclePosition()
+{
+    if (isOnVehicle())
+        return mMovementController->getVehicleState()->getPosition(mPosTypeUsed);
+    else
+        return mVehicleConnection->getVehicleState()->getPosition(mPosTypeUsed);
+}
+
+void FollowPoint::updateState()
+{
+    QLineF egoVehicleToFollowPointLine(QPointF(0,0), mCurrentFollowPointInVehicleFrame.getPoint());
+
+    switch (mStmState) {
+    case FollowPointSTMstates::STOP:
+        break;
```

Comment:
```suggestion
    case FollowPointSTMstates::NONE:
        qDebug() << "WARNING: FollowPoint running uninitialized statemachine.";
        break;
```
To make sure you get a warning when the STM state is executed, which to my understanding should never happen. Would call it NONE for consistency with waypointfollower